site/src/pages/TemplateVersionEditorPage/TemplateVersionEditorPage.tsx

```diff
@@ -1,115 +1,359 @@
+import { useMutation, useQuery, useQueryClient } from "react-query";
+import {
+  createTemplateVersion,
+  resources,
+  templateByName,
+  templateVersionByName,
+  templateVersionVariables,
+} from "api/queries/templates";
+import { file, uploadFile } from "api/queries/files";
+import { TarReader, TarWriter } from "utils/tar";
+import { FileTree, traverse } from "utils/filetree";
+import {
+  createTemplateVersionFileTree,
+  isAllowedFile,
+} from "utils/templateVersion";
+import {
+  patchTemplateVersion,
+  updateActiveTemplateVersion,
+  watchBuildLogsByTemplateVersionId,
+} from "api/api";
+import {
+  PatchTemplateVersionRequest,
+  ProvisionerJobLog,
+  TemplateVersion,
+} from "api/typesGenerated";
 
 type Params = {
   version: string;
   template: string;
 };
 
 export const TemplateVersionEditorPage: FC = () => {
+  const queryClient = useQueryClient();
   const navigate = useNavigate();
-  const { version: versionName, template: templateName } =
+  const { version: initialVersionName, template: templateName } =
     useParams() as Params;
   const orgId = useOrganizationId();
-  const [editorState, sendEvent] = useMachine(templateVersionEditorMachine, {
-    context: { orgId },
+  const [currentVersionName, setCurrentVersionName] =
+    useState(initialVersionName);
+  const templateQuery = useQuery(templateByName(orgId, templateName));
+  const templateVersionOptions = templateVersionByName(
+    orgId,
+    templateName,
+    currentVersionName,
+  );
+  const templateVersionQuery = useQuery({
+    ...templateVersionOptions,
+    keepPreviousData: true,
   });
-  const { isSuccess, data } = useTemplateVersionData(
-    {
-      orgId,
-      templateName,
-      versionName,
-    },
-    {
-      onSuccess(data) {
-        sendEvent({ type: "INITIALIZE", tarReader: data.tarReader });
-      },
-    },
+  const uploadFileMutation = useMutation(uploadFile());
+  const createTemplateVersionMutation = useMutation(
+    createTemplateVersion(orgId),
   );
+  const resourcesQuery = useQuery({
+    ...resources(templateVersionQuery.data?.id ?? ""),
+    enabled: templateVersionQuery.data?.job.status === "succeeded",
+  });
+  const { logs, setLogs } = useVersionLogs(templateVersionQuery.data, {
+    onDone: templateVersionQuery.refetch,
+  });
+  const { fileTree, tarFile } = useFileTree(templateVersionQuery.data);
+  const {
+    missingVariables,
+    setIsMissingVariablesDialogOpen,
+    isMissingVariablesDialogOpen,
+  } = useMissingVariables(templateVersionQuery.data);
+
+  // Handle template publishing
+  const [isPublishingDialogOpen, setIsPublishingDialogOpen] = useState(false);
+  const publishVersionMutation = useMutation({
+    mutationFn: publishVersion,
+  });
+  const [lastSuccessfulPublishedVersion, setLastSuccessfulPublishedVersion] =
+    useState<TemplateVersion>();
+
+  // Optimistically update the template version data job status to make the
+  // build action feels faster
+  const onBuildStart = () => {
```

Comment:
Could some of this functionality be done via `useMutation`'s `onMutate` and `onSuccess` callbacks? `onMutate` also lets you return out a cleanup function

Reply:
I don't think so because this is not related to a specific mutation.